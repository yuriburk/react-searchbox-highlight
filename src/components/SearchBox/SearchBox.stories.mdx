import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';

import SearchBox, { ISearchBoxProps } from '.';

<Meta title="Components/SearchBox" component={SearchBox} />

# SearchBox

This is the `SearchBox`

```jsx
import { SearchBox } from 'react-searchbox';
```

<Preview withSource="open">
  <Story name="SearchBox">
    <SearchBox
      onChange={(value) => console.log('change')}
      items={[
        { label: 'teste', value: 'teste' },
        { label: 'teste2', value: 'teste2' },
        { label: 'teste2', value: 'teste2' },
        { label: 'teste2', value: 'teste2' },
        { label: 'teste2', value: 'teste2' },
        { label: 'teste2', value: 'teste2' },
      ]}
      onItemClick={(value) => console.log('click')}
    />
  </Story>
</Preview>

## Props

<Props of={SearchBox} />
