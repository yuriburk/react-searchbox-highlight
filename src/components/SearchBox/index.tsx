import React, {
  useRef,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react';
import debounce from 'lodash.debounce';

import Highlighter from '../Highlighter';
import {
  Container,
  Card,
  SearchIcon,
  Input,
  LoadingIcon,
  CloseIcon,
  ItemsContainer,
  Item,
} from './styles';

export interface ISearchBoxProps {
  items: { label: string; value: unknown }[];
  onChange(value: string): void;
  onItemClick(value: unknown): void;
  debounceTime?: number;
  placeholder?: string;
  highlighterStyle?: Record<string, unknown>;
}

const SearchBox = ({
  items,
  onChange,
  onItemClick,
  debounceTime = 200,
  placeholder = 'Search...',
  highlighterStyle = {
    color: '#fff',
    backgroundColor: '#9999ff',
  },
}: ISearchBoxProps) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const isMounted = useRef(true);

  const [isLoading, setIsLoading] = useState(false);
  const [searchWords, setSearchWords] = useState<string[]>([]);

  const handleDebouncedChange = useMemo(
    () =>
      debounce((value: string) => {
        onChange(value);

        if (isMounted.current) {
          setIsLoading(false);
        }
      }, debounceTime),
    [onChange, debounceTime]
  );

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  const handleClearInput = useCallback(() => {
    if (inputRef.current) {
      inputRef.current.value = '';
    }

    setIsLoading(false);
    setSearchWords([]);
  }, []);

  const handleOnChange = useCallback(
    (event) => {
      event.persist();
      setIsLoading(true);
      const { value } = event.target;

      if (value.length > 0) {
        setSearchWords(value.split(' '));
        handleDebouncedChange(value);
      } else {
        handleClearInput();
      }
    },
    [handleDebouncedChange, handleClearInput]
  );

  const handleOnItemClick = useCallback(
    (value) => {
      handleClearInput();
      onItemClick(value);
    },
    [handleClearInput, onItemClick]
  );

  return (
    <Container>
      <Card>
        <SearchIcon />
        <Input
          ref={inputRef}
          type="text"
          placeholder={placeholder}
          onChange={handleOnChange}
        />
        {isLoading ? (
          <LoadingIcon />
        ) : (
          searchWords.length > 0 && (
            <CloseIcon onClick={() => handleClearInput()} />
          )
        )}
      </Card>
      {items?.length > 0 && searchWords.length > 0 && (
        <ItemsContainer>
          {items.map((item, index) => (
            <Item key={index} onClick={() => handleOnItemClick(item.value)}>
              <Highlighter
                searchWords={searchWords}
                text={item.label}
                style={highlighterStyle}
              />
            </Item>
          ))}
        </ItemsContainer>
      )}
    </Container>
  );
};

export default SearchBox;
