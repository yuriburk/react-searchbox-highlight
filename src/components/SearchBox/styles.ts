import styled from 'styled-components';
import { FiSearch, FiX } from 'react-icons/fi';

export const Container = styled.div`
  position: relative;
`;

export const Card = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;

  padding: 10px;
  box-sizing: border-box;
  border: 2px solid;
  border-radius: 4px;
  border-color: #d2d2d2;
  background-color: #fff;
  box-shadow: 0px 1px 4px 0px rgba(240, 242, 245, 1);
`;

export const SearchIcon = styled(FiSearch)`
  color: #d2d2d2;
  width: 24px;
  height: 24px;
  margin-right: 4px;
`;

export const Input = styled.input`
  border: 0;
  outline: 0;
  width: 100%;
  font-size: 18px;
`;

export const LoadingIcon = styled.div`
  display: flex;
  align-items: center;

  @keyframes spin {
    to {
      transform: rotate(360deg);
    }
  }

  border: 4px solid rgba(0, 0, 0, 0.1);
  border-top: 4px solid #d2d2d2;
  border-radius: 50%;
  width: 14px;
  height: 14px;
  animation: spin 2s ease-in-out infinite;
`;

export const CloseIcon = styled(FiX)`
  cursor: pointer;
  color: #d2d2d2;
  width: 24px;
  height: 24px;
  margin-right: 4px;
`;

export const ItemsContainer = styled(Card)`
  display: flex;
  flex-direction: column;
  align-items: flex-start;

  position: absolute;
  overflow: auto;
  max-height: 289px;
  width: 100%;
  margin: 0;
  border-top: 0;
  z-index: 1;

  div + div {
    border-top: 2px solid;
    border-color: #d2d2d2;
  }
`;

export const Item = styled.div`
  cursor: pointer;
  position: relative;
  width: 100%;
  padding: 14px;
  color: #000;
`;
