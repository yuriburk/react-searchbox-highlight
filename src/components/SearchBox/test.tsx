import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import SearchBox from '.';

const mockedItems = [
  { label: 'John One - 1', value: { name: 'John One' } },
  { label: 'John Duo - 2', value: { name: 'John Duo' } },
];
const mockedOnChange = jest.fn();
const mockedItemClick = jest.fn();

describe('<SearchBox />', () => {
  beforeEach(() => jest.clearAllMocks());

  it('should be able to take a snapshot', async () => {
    const { container, getByRole } = render(
      <SearchBox
        items={mockedItems}
        onChange={mockedOnChange}
        onItemClick={mockedItemClick}
        placeholder="Search here"
      />
    );

    await userEvent.type(getByRole('textbox'), 'John');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should be able to handle change event in input text', async () => {
    const { getByRole } = render(
      <SearchBox
        items={mockedItems}
        onChange={mockedOnChange}
        onItemClick={mockedItemClick}
        placeholder="Search here"
      />
    );

    await userEvent.type(getByRole('textbox'), 'John');

    await waitFor(() => expect(mockedOnChange).toHaveBeenCalled());
  });

  it('should be able to handle blur event in input blur', async () => {
    const { getByRole, queryAllByText } = render(
      <SearchBox
        items={mockedItems}
        onChange={mockedOnChange}
        onItemClick={mockedItemClick}
        placeholder="Search here"
      />
    );

    const inputElement = getByRole('textbox');
    userEvent.click(inputElement);
    expect(inputElement).toHaveFocus();
    await userEvent.type(inputElement, 'John');

    const itemsElements = queryAllByText('John');
    await waitFor(() => {
      expect(itemsElements[0]).toBeInTheDocument();
      expect(itemsElements[1]).toBeInTheDocument();
    });

    const cardElement = itemsElements[0];
    userEvent.click(cardElement);

    expect(inputElement).not.toHaveFocus();
    await waitFor(() => {
      expect(itemsElements[0]).not.toBeInTheDocument();
      expect(itemsElements[1]).not.toBeInTheDocument();
    });
  });
});
