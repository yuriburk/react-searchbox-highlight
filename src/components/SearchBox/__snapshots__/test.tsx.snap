// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox /> should be able to take a snapshot 1`] = `
<div
  class="sc-gtssRu lgvXzJ"
>
  <div
    class="sc-dlnjPT khItPs"
  >
    <svg
      class="sc-hKFyIo eXrevn"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="11"
        cy="11"
        r="8"
      />
      <line
        x1="21"
        x2="16.65"
        y1="21"
        y2="16.65"
      />
    </svg>
    <input
      class="sc-eCApGN JePFZ"
      placeholder="Search here"
      type="text"
    />
    <div
      class="sc-jSFkmK eCZsJq"
    />
  </div>
  <div
    class="sc-dlnjPT sc-iCoHVE khItPs jpqSiK"
  >
    <div
      class="sc-fujyUd fzFLqJ"
    >
      <span
        class="sc-bdnylx"
      >
        <mark
          class=" "
          style="color: rgb(255, 255, 255); background-color: rgb(153, 153, 255);"
        >
          John
        </mark>
        <span
          class=""
        >
           One - 1
        </span>
      </span>
    </div>
    <div
      class="sc-fujyUd fzFLqJ"
    >
      <span
        class="sc-bdnylx"
      >
        <mark
          class=" "
          style="color: rgb(255, 255, 255); background-color: rgb(153, 153, 255);"
        >
          John
        </mark>
        <span
          class=""
        >
           Duo - 2
        </span>
      </span>
    </div>
  </div>
</div>
`;
